kv/kvnemesis: TestKVNemesisMultiNode_BufferedWritesNoPipelining failed

kv/kvnemesis.TestKVNemesisMultiNode_BufferedWritesNoPipelining [failed](https://tanzanite.cluster.engflow.com/invocations/default/2ecc89bf-254f-4634-866a-f1c6cbf73be0?testReportRun=6&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZuZW1lc2lzOmt2bmVtZXNpc190ZXN0) on master @ [a53b28b8a8d1f1f1544f036bfbdb0952e5371388](https://github.com/cockroachdb/cockroach/commits/a53b28b8a8d1f1f1544f036bfbdb0952e5371388):


```
          txn.Del(ctx, tk(17701988292606188248) /* @s32 */) // <nil>
          {
            b := &kv.Batch{}
            b.GetForShareGuaranteedDurability(tk(5527626609645510665)) // (v31, <nil>)
            b.GetForUpdateGuaranteedDurability(tk(20852549119673753)) // (<nil>, <nil>)
            b.ReverseScanForUpdateGuaranteedDurability(tk(676529557188221280), tk(13769969139351669691)) // (/Table/100/"8c4104827135056f":v22, /Table/100/"85cdc13c13db1457":v23, /Table/100/"62249102519367b0":v17, /Table/100/"5be00935c522674b":v10, /Table/100/"4cb6131d66eae009":v31, /Table/100/"323c40d198c17fdb":v14, /Table/100/"2ddfed6dda906f12":v14, /Table/100/"2d0f77321b93c624":v14, /Table/100/"28f83443faed1cb5":v14, /Table/100/"27d2ecc82d338d1b":v14, /Table/100/"1fa2cbe69363763a":v14, /Table/100/"18fb8de662f8208c":v14, /Table/100/"0daefc057d04b0a4":v14, /Table/100/"0c099a4cb7cd283a":v14, <nil>)
            txn.Run(ctx, b) // @1750829877.735968317,0 <nil>
          }
          {
            b := &kv.Batch{}
            b.ReverseScanForUpdate(tk(10517747571112718509), tk(10806739655496939535)) // <nil>
            b.Del(tk(14299019860527645387) /* @s33 */) // <nil>
            b.ReverseScanForUpdate(tk(2284444448191326710), tk(14689524263264116362)) // (/Table/100/"8c4104827135056f":v22, /Table/100/"85cdc13c13db1457":v23, /Table/100/"62249102519367b0":v17, /Table/100/"5be00935c522674b":v10, /Table/100/"4cb6131d66eae009":v31, /Table/100/"323c40d198c17fdb":v14, /Table/100/"2ddfed6dda906f12":v14, /Table/100/"2d0f77321b93c624":v14, /Table/100/"28f83443faed1cb5":v14, /Table/100/"27d2ecc82d338d1b":v14, <nil>)
            txn.Run(ctx, b) // @1750829877.735968317,0 <nil>
          }
          txn.ScanForShare(ctx, tk(7154085541646664476), tk(9661629885016068844), 0) // @1750829877.735968317,0 (/Table/100/"85cdc13c13db1457":v23, <nil>)
          txn.GetForShare(ctx, tk(13711426494513486876)) // @1750829877.735968317,0 (<nil>, <nil>)
          txn.GetForShare(ctx, tk(311629518298168857)) // @1750829877.735968317,0 (<nil>, <nil>)
          txn.Put(ctx, tk(17701988292606188248), sv(34)) // <nil>
          txn.CreateSavepoint(ctx, 13) // <nil>
          txn.ScanForUpdateSkipLockedGuaranteedDurability(ctx, tk(733569472353398386), tk(8655174367651819296), 0) // @1750829877.735968317,0 (/Table/100/"0c099a4cb7cd283a":v14, /Table/100/"0daefc057d04b0a4":v14, /Table/100/"18fb8de662f8208c":v14, /Table/100/"1fa2cbe69363763a":v14, /Table/100/"27d2ecc82d338d1b":v14, /Table/100/"28f83443faed1cb5":v14, /Table/100/"2d0f77321b93c624":v14, /Table/100/"2ddfed6dda906f12":v14, /Table/100/"323c40d198c17fdb":v14, /Table/100/"4cb6131d66eae009":v31, /Table/100/"5be00935c522674b":v10, /Table/100/"62249102519367b0":v17, <nil>)
          b := &kv.Batch{}
          b.GetForShare(tk(17337079236026049797)) // cannot acquire lock with strength Exclusive at seq number 0, already held at higher seq number 1
          txn.CommitInBatch(ctx, b) // cannot acquire lock with strength Exclusive at seq number 0, already held at higher seq number 1
          return nil
        }) // cannot acquire lock with strength Exclusive at seq number 0, already held at higher seq number 1: cannot acquire lock with strength Exclusive at seq number 0, already held at higher seq number 1
    kvnemesis.go:173: error applying x.GetForShare(ctx, tk(17337079236026049797)) // cannot acquire lock with strength Exclusive at seq number 0, already held at higher seq number 1: cannot acquire lock with strength Exclusive at seq number 0, already held at higher seq number 1
    kvnemesis.go:208: failures(verbose): /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_6_of_10/test.outputs/kvnemesis2784980859/failures
        repro steps: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_6_of_10/test.outputs/kvnemesis2784980859/repro.go
        rangefeed KVs: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_6_of_10/test.outputs/kvnemesis2784980859/kvs-rangefeed.txt
        scan KVs: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_6_of_10/test.outputs/kvnemesis2784980859/kvs-scan.txt
    kvnemesis_test.go:459: [0] proposed: 2072
    kvnemesis_test.go:461: [0] reproposed unchanged: 75
    kvnemesis_test.go:463: [0] reproposed with new LAI: 2
    kvnemesis_test.go:459: [1] proposed: 200
    kvnemesis_test.go:461: [1] reproposed unchanged: 2
    kvnemesis_test.go:463: [1] reproposed with new LAI: 4
    kvnemesis_test.go:459: [2] proposed: 0
    kvnemesis_test.go:461: [2] reproposed unchanged: 0
    kvnemesis_test.go:463: [2] reproposed with new LAI: 0
    kvnemesis_test.go:468: 
        	Error Trace:	pkg/kv/kvnemesis/kvnemesis_test.go:468
        	            				pkg/kv/kvnemesis/kvnemesis_test.go:363
        	Error:      	Should be zero, but was 2
        	Test:       	TestKVNemesisMultiNode_BufferedWritesNoPipelining
        	Messages:   	kvnemesis detected failures
    dump_raft_logs.go:32: dumping raft logs to /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_6_of_10/test.outputs/kvnemesis2784980859
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestKVNemesisMultiNode_BufferedWritesNoPipelining2851873523
--- FAIL: TestKVNemesisMultiNode_BufferedWritesNoPipelining (7.24s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=6</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode_BufferedWritesNoPipelining.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-51803

--- a/pkg/kv/kvserver/batcheval/cmd_get.go
+++ b/pkg/kv/kvserver/batcheval/cmd_get.go
@@ -120,8 +120,13 @@ func Get(
 	shouldLockKey := getRes.Value != nil || args.LockNonExisting
 	var res result.Result
 	if args.KeyLockingStrength != lock.None && shouldLockKey {
+		// ExpectExclusionSince is used by callers (namely, txnWriteBuffers) that
+		// are likely to be sending replicated, locking Get requests at sequence
+		// numbers corresponding to unreplicated locks taken earlier in the
+		// transaction. In this case, sequence number regression is not unexpected.
+		allowSequenceNumberRegression := args.ExpectExclusionSince.IsSet()
 		acq, err := acquireLockOnKey(ctx, readWriter, h.Txn, args.KeyLockingStrength,
-			args.KeyLockingDurability, args.Key, cArgs.Stats, cArgs.EvalCtx.ClusterSettings())
+			args.KeyLockingDurability, args.Key, cArgs.Stats, cArgs.EvalCtx.ClusterSettings(), allowSequenceNumberRegression)
 		if err != nil {
 			return result.Result{}, err
 		}

--- a/pkg/kv/kvserver/batcheval/lock.go
+++ b/pkg/kv/kvserver/batcheval/lock.go
@@ -137,7 +137,7 @@ func acquireLocksOnKeys(
 	case kvpb.BATCH_RESPONSE:
 		err := storage.MVCCScanDecodeKeyValues(scanRes.KVData, func(key storage.MVCCKey, _ []byte) error {
 			k := copyKey(key.Key)
-			acq, err := acquireLockOnKey(ctx, readWriter, txn, str, dur, k, ms, settings)
+			acq, err := acquireLockOnKey(ctx, readWriter, txn, str, dur, k, ms, settings, false /* 	allowSequenceNumberRegression */)
 			if err != nil {
 				return err
 			}
@@ -153,7 +153,7 @@ func acquireLocksOnKeys(
 	case kvpb.KEY_VALUES:
 		for _, row := range scanRes.KVs {
 			k := copyKey(row.Key)
-			acq, err := acquireLockOnKey(ctx, readWriter, txn, str, dur, k, ms, settings)
+			acq, err := acquireLockOnKey(ctx, readWriter, txn, str, dur, k, ms, settings, false /* 	allowSequenceNumberRegression */)
 			if err != nil {
 				return nil, err
 			}
@@ -198,6 +198,7 @@ func acquireLockOnKey(
 	key roachpb.Key,
 	ms *enginepb.MVCCStats,
 	settings *cluster.Settings,
+	allowSequenceNumberRegression bool,
 ) (roachpb.LockAcquisition, error) {
 	maxLockConflicts := storage.MaxConflictsPerLockConflictError.Get(&settings.SV)
 	targetLockConflictBytes := storage.TargetBytesPerLockConflictError.Get(&settings.SV)
@@ -233,7 +234,8 @@ func acquireLockOnKey(
 		// conflicts with un-contended replicated locks -- we need to do so before
 		// we can acquire our own replicated lock; do that now, and also acquire
 		// the replicated lock if no conflicts are found.
-		if err := storage.MVCCAcquireLock(ctx, readWriter, &txn.TxnMeta, txn.IgnoredSeqNums, str, key, ms, maxLockConflicts, targetLockConflictBytes, false); err != nil {
+		if err := storage.MVCCAcquireLock(ctx, readWriter, &txn.TxnMeta, txn.IgnoredSeqNums, str,
+			key, ms, maxLockConflicts, targetLockConflictBytes, allowSequenceNumberRegression); err != nil {
 			return roachpb.LockAcquisition{}, err
 		}
 	default:

